R/detect_column.r

```diff
@@ -0,0 +1,354 @@
+  if(domain_spec=="CO"){
+    data("CO")
+    domain_spec <- force(CO)
+  } else if (domain_spec=="DM"){
+    data("DM")
+    domain_spec <- force(DM)
+  } else if (domain_spec=="SE"){
+    data("SE")
+    domain_spec <- force(SE)
+  } else if (domain_spec=="SM"){
+    data("SM")
+    domain_spec <- force(SM)
+  } else if (domain_spec=="SV"){
+    data("SV")
+    domain_spec <- force(SV)
+  }
+  permitted_cols <- domain_spec$`Variable Name`[domain_spec$Core %in% "Perm"]
+  missing_permitted <- setdiff(permitted_cols, names(data))
+  if (length(missing_permitted) > 0) {
+    return(NULL)
+  } 
+}
+
+# validate_permitted_column_names(dat, domain_spec=domain_specs$DM)
+
+#' Check if extra columns exist, give an warning if they do exist
+#' 
+#' @param data The dataset whose columns to check.
+#' @param domain_spec The domain specification ("CO", "DM", "SE", "SM", "SV") for checking columns.
+#' @return If extra columns exist, give an warning and indicate which extra columns exist.
+#' @export
+validate_extra_column_names <- function(data, domain_spec) {
+  if(domain_spec=="CO"){
+    data("CO")
+    domain_spec <- force(CO)
+  } else if (domain_spec=="DM"){
+    data("DM")
+    domain_spec <- force(DM)
+  } else if (domain_spec=="SE"){
+    data("SE")
+    domain_spec <- force(SE)
+  } else if (domain_spec=="SM"){
+    data("SM")
+    domain_spec <- force(SM)
+  } else if (domain_spec=="SV"){
+    data("SV")
+    domain_spec <- force(SV)
+  }
+  all_cols <- domain_spec$`Variable Name`
+  extra_cols <- setdiff(names(data),all_cols)
+  if (length(extra_cols) > 0) {
+    warning(
+      "The following extra columns do exist: ",
+      paste(extra_cols, collapse=", ")
+    )
+  } 
+}
+
+# validate_extra_column_names(dat, domain_spec=domain_specs$DM)
+
+#' Check to see if columns have the expected class (numeric or character), if not give a warning and convert to one or the other. Also, check to see if columns contain NA, give an error if they do.
+#' 
+#' @param data The dataset whose columns to check.
+#' @param domain_spec The domain specification ("CO", "DM", "SE", "SM", "SV") for checking columns.
+#' @return If columns are numeric but should be character, give a warning and indicate which are those columns. If columns are character but should be numeric, give a warning and indicate those columns. If columns have NA, give a warning and indicate which are those columns. 
+#' @export
+validate_column_class <- function(data,domain_spec){
+  if(domain_spec=="CO"){
+    data("CO")
+    domain_spec <- force(CO)
+  } else if (domain_spec=="DM"){
+    data("DM")
+    domain_spec <- force(DM)
+  } else if (domain_spec=="SE"){
+    data("SE")
+    domain_spec <- force(SE)
+  } else if (domain_spec=="SM"){
+    data("SM")
+    domain_spec <- force(SM)
+  } else if (domain_spec=="SV"){
+    data("SV")
+    domain_spec <- force(SV)
+  }
+  dat1 <- lapply(data,class)
+  dat1 <- data.frame(Variable = names(unlist(dat1)), Type = unlist(dat1))
```

Comment:
Classes are not all scalars.  This will cause an error (or unexpected bug) for any non-scalar classes.

For example all date-time classes have two names:

``` r
class(as.POSIXct("2020-04-11 12:13"))
#> [1] "POSIXct" "POSIXt"
```

<sup>Created on 2020-04-11 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>